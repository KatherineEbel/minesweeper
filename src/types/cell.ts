export type Cell = 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12

export const CellState: Record<string, Cell> = {
  empty: 0,
  bomb: 9,
  hidden: 10,
  flag: 11,
  weakFlag: 12,
}
