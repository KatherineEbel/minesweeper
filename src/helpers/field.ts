import { Cell, CellState, Coordinates, Field, Neighbors } from '../types'

export const isActive = (cell: Cell): boolean => {
  return [CellState.hidden, CellState.flag, CellState.weakFlag].includes(cell)
}

export const exists = ([row, col]: Coordinates, field: Field): boolean => {

  return field[row] !== undefined && field[row][col] !== undefined
}

export const getNeighbors = ([row, col]: Coordinates): Neighbors => {
  return {
    top: [row - 1, col],
    topRight: [row - 1, col + 1],
    right: [row, col + 1],
    rightBottom: [row + 1, col + 1],
    bottom: [row + 1, col],
    bottomLeft: [row + 1, col - 1],
    left: [row, col - 1],
    leftTop: [row - 1, col - 1],
  }
}

export const MineSweeper = {
  buildEmpty: (size: number, state: Cell = CellState.empty): Field => {
    return new Array(size).fill(null).map(() => new Array(size).fill(state))
  },
  buildField: (size: number, probability: number): Field => {
    if (probability < 0 || probability > 1)
      throw new Error('Density must be between 0 and 1')

    let unprocessedCells = size * size
    let restCellsWithBombs = unprocessedCells * probability
    const mineField = MineSweeper.buildEmpty(size)

    for (let i = 0; i < size; i++) {
      for (let j = 0; j < size; j++) {
        if (restCellsWithBombs === 0) return mineField
        if (restCellsWithBombs / unprocessedCells > Math.random()) {
          mineField[i][j] = CellState.bomb
          MineSweeper.incrementNeighbors([i, j], mineField)
          restCellsWithBombs--
        }
        unprocessedCells--
      }
    }
    return mineField
  },
  incrementNeighbors: (coords: Coordinates, field: Field): Field => {
    const neighbors = getNeighbors(coords)
    Object.values(neighbors).forEach((coords) => {
      const [row, col] = coords
      if (exists(coords, field)) {
        const cell: Cell = field[row][col]
        if (cell < CellState.bomb - 1) {
          field[row][col] += 1
        }
      }
    })
    return field
  },
}
