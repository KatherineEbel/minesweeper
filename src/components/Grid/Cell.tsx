import tw, {styled} from 'twin.macro'
import {Cell as CellType, CellState } from '../../types/cell'
import React from 'react'
import {isActive, Coordinates} from '../../helpers/field'
import {useMouseDown} from '../../hooks/useMouseDown'

const transparent = 'rgba(0,0,0.0)'
const colors: { [key in CellType ]: string} = {
  0: '#0f172a',
  1: '#bae6fd',
  2: '#99f6e4',
  3: '#a78bfa',
  4: '#f0abfc',
  5: '#bef264',
  6: '#fb923c',
  7: '#fbbf24',
  8: '#fda4af',
  9: 'transparent',
  10: 'transparent',
  11: 'transparent',
  12: 'transparent',
}

export interface CellProps {
  type: CellType
  coords: Coordinates
  mouseDown?: boolean,
  onClick?: (coords: Coordinates) => void
  onContextMenu: (coords: Coordinates) => void
  onMouseUp?: () => void
  onMouseDown?: () => void
}

const CellContainer = tw.div`
flex items-center justify-center cursor-pointer w-8 h-8 bg-slate-700 border border-slate-600
`

type HiddenCellProps = Pick<CellProps, 'mouseDown'>

const HiddenCell = styled(CellContainer)<HiddenCellProps>`
  ${tw`select-none hover:bg-slate-600 transition-all duration-300 text-transparent`}
  box-shadow: -3px 3px rgba(0, 0, 0, 0.7), -2px 2px rgba(0, 0, 0, 0.7), -1px 1px rgba(0, 0, 0, 0.7);
  border: 1px solid ${({mouseDown = false}) => mouseDown ? 'transparent' : 'rgba(0,0,0,0.8)'};

  &:active {
    box-shadow: -1px 1px rgba(0, 0, 0, 0.7), -2px 2px rgba(0, 0, 0, 0.7), -1px 1px rgba(0, 0, 0, 0.7);
  }
`


const OpenCell = styled(CellContainer)<Pick<CellProps, 'type'>>`
  ${tw`cursor-default border-slate-500`}
  color: ${({ type }) => colors[type as CellType] ?? transparent};
`

const EmptyCell = styled(OpenCell)`${tw`text-transparent`}`

const StyledFlag = styled.div`
  width: 0;
  height: 0;
  border-top: .5rem solid transparent;
  border-bottom: .5rem solid transparent;
`

const Flag = styled(StyledFlag)`
  border-left: .5rem solid #ec433c;
`

const WeakFlag = styled(StyledFlag)`
  border-left: .5rem solid #fcd34d
`

const Bomb = tw.div`
  rounded-full w-4 h-4 bg-slate-900
`

const BombFrame = tw(OpenCell)`bg-rose-600`

const Cell = ({type, coords, ...rest}: CellProps) => {
  const [mouseDown, onMouseUp, onMouseDown] = useMouseDown()
  const isActiveCell = isActive(type)

  const onClick = () => rest.onClick && rest.onClick(coords)

  const onContextMenu = (el: React.MouseEvent<HTMLElement>) => {
    el.preventDefault()
    if (isActiveCell) rest.onContextMenu(coords)
  }

  const commonProps = {
    className: `cell-${coords.join('-')}`,
    onContextMenu,
    type,
    role: 'cell'
  }

  const activeProps = {
    ...commonProps,
    mouseDown,
    onClick,
    onMouseUp,
    onMouseDown,
  }


  switch (type) {
    case CellState.empty:
      return <EmptyCell {...commonProps}>{type}</EmptyCell>
    case CellState.bomb:
      return <BombFrame {...commonProps}>
        <Bomb/>
      </BombFrame>
    case CellState.hidden:
      return <HiddenCell {...activeProps}>{type}</HiddenCell>
    case CellState.flag:
      return <HiddenCell {...activeProps}>
        <Flag>
          {type}
        </Flag>
      </HiddenCell>
    case CellState.weakFlag:
      return <HiddenCell {...activeProps}>
        <WeakFlag>
          {type}
        </WeakFlag>
      </HiddenCell>
    default:
      return <OpenCell {...commonProps}>{type}</OpenCell>
  }
}

export default Cell