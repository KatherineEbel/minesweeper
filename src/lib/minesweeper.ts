import {Cell, CellState} from './cell'
import {Coordinates, exists, Field, getNeighbors, isMine, isEmpty, isFlag, isHidden} from './helpers/field'

export const MineSweeper = {
  buildEmpty: (size: number, state: Cell = CellState.empty): Field => {
    return new Array(size).fill(null).map(() => new Array(size).fill(state))
  },
  buildField: (size: number, probability: number): Field => {
    if (probability < 0 || probability > 1)
      throw new Error('Density must be between 0 and 1')

    let unprocessedCells = size * size
    let restCellsWithBombs = unprocessedCells * probability
    const mineField = MineSweeper.buildEmpty(size)

    for (let i = 0; i < size; i++) {
      for (let j = 0; j < size; j++) {
        if (restCellsWithBombs === 0) return mineField
        if (restCellsWithBombs / unprocessedCells > Math.random()) {
          mineField[i][j] = CellState.mine
          MineSweeper.incrementNeighbors([i, j], mineField)
          restCellsWithBombs--
        }
        unprocessedCells--
      }
    }
    return mineField
  },
  incrementNeighbors: (coords: Coordinates, field: Field): Field => {
    const neighbors = getNeighbors(coords)
    Object.values(neighbors).forEach((coords) => {
      const [row, col] = coords
      if (exists(coords, field)) {
        const cell: Cell = field[row][col]
        if (cell < CellState.mine - 1) {
          field[row][col] += 1
        }
      }
    })
    return field
  },
  /**
   * Given coordinates, playerField and gameField open cell at coordinates
   * Throws error if opened cell is a mine
   * Returns updated playerField
   * @param row
   * @param col
   * @param playerField
   * @param gameField
   * @returns Field
   */
  openCell: ([row, col]: Coordinates, playerField: Field, gameField: Field): Field => {
    const gameCell = gameField[row][col]
    playerField[row][col] = gameCell
    if (isMine(gameCell)) throw new Error('Game Over')
    if (isEmpty(gameCell)) {
      const neighbors = getNeighbors([row, col])
      Object.values(neighbors).forEach(neighbor => {
        if (exists(neighbor, gameField)) {
          const [r, c] = neighbor
          const gameCell = gameField[r][c]
          const playerCell = playerField[r][c]
          if (isEmpty(gameCell) && isHidden(playerCell)) {
            playerField = MineSweeper.openCell(neighbor, playerField, gameField)
          }
          if (gameCell < CellState.mine) {
            playerField[r][c] = gameCell
          }
        }
      })
    }
    return playerField
  },
  /**
   * Sets flag in field or throws error if max flags reached
   * @param coords
   * @param playerField
   * @param flagCount
   * @param mineCount
   * @returns [Field, number]
   *
   */
  setFlag: (coords: Coordinates, playerField: Field, flagCount: number, mineCount: number): [Field, number] => {
    const [row, col] = coords
    const cell = playerField[row][col]
    let flagDiff = 0
    const {flag, weakFlag, hidden} = CellState
    switch (cell) {
      case flag:
        playerField[row][col] = weakFlag
        break
      case weakFlag:
        flagDiff--
        playerField[row][col] = hidden
        break
      case hidden:
        if (flagCount === mineCount) throw new Error('Remove a flag to add a new one')
        playerField[row][col] = flag
        flagDiff++
        break
      default:
        return [playerField, flagDiff]
    }
    return [playerField, flagDiff]
  },
  /**
   * Detects if game is solved dependent on playerField and gameField
   * @param playerField
   * @param gameField
   * @returns boolean
   */
  detectSolved: (playerField: Field, gameField: Field): boolean => {
    const {flag, weakFlag} = CellState
    let [mineCount, flagCount, detectedMines, hiddenCount] = [0,0,0,0];
    for (const row of gameField.keys()) {
      for (const col of gameField[row].keys()) {
        const gameCell = gameField[row][col]
        const playerCell = playerField[row][col]
        const mineFound = isMine(gameCell)
        if (mineFound) {
          mineCount++
          if (isFlag(playerCell)) detectedMines++
        }
        if (isHidden(playerCell)) {
          hiddenCount++
        }
        if ([flag, weakFlag].includes(playerCell)) {
          flagCount++
        }
      }
    }
    return mineCount === detectedMines &&  hiddenCount === 0
  },
}
