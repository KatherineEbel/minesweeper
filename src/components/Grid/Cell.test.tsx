import Cell, {areEqual} from 'components/Grid/Cell'
import {Cell as CellType} from 'lib/cell'
import {createEvent, fireEvent, render } from '@testing-library/react'
import {isActive, Coordinates} from 'lib/helpers/field'

describe('Cell Component', function () {
  const coords: Coordinates = [1, 1]
  const props = {
    coords,
    onClick: jest.fn(),
    onContextMenu: jest.fn()
  }
  const cells = Array.from({length: 13}, (_, k) => ({cell: k as CellType, prevented: true}))
  test.each(cells)('check prevent default for context menu', ({cell, prevented}) => {
    const props = { coords, onClick: jest.fn(), onContextMenu: jest.fn() }
    const {container} = render(<Cell cellType={cell } {...props}/>)
    const target = container.querySelector(`.cell-${coords.join('-')}`)!
    const contextMenuEvent = createEvent.contextMenu(target)
    fireEvent(target, contextMenuEvent)
    expect(contextMenuEvent.defaultPrevented).toBe(prevented)
  })


  test.each(cells.filter(({cell}) => isActive(cell)))('onClick and onContextMenu handlers called for active cells', ({cell }) => {
    const props = { coords, onClick: jest.fn(), onContextMenu: jest.fn() }
    const {container } = render(<Cell cellType={cell } {...props}/>)
    const target = container.querySelector(`.cell-${coords.join('-')}`)!
    if (isActive(cell)) {
      expect(target).toHaveStyle(`border: 1px solid rgba(0,0,0,0.8)`)
      fireEvent.mouseDown(target)
      expect(target).toHaveStyle(`border: 1px solid transparent`)
    }
    fireEvent.click(target)
    fireEvent.contextMenu(target)
    expect(props.onClick).toBeCalled()
    expect(props.onContextMenu).toBeCalled()
  })

  test('areEqual', () => {
    const prevProps = {
      ...props,
      cellType: 0 as CellType,
    }

    expect(areEqual(prevProps, {...prevProps})).toBe(true)

    expect(areEqual(prevProps, {...prevProps, coords: [2, 1]})).toBe(false)
    expect(areEqual(prevProps, {...prevProps, coords: [1, 2]})).toBe(false)
    expect(areEqual(prevProps, {...prevProps, coords: [2, 2]})).toBe(false)
    expect(areEqual(prevProps, {...prevProps, coords: [1, 0]})).toBe(false)
  })
})