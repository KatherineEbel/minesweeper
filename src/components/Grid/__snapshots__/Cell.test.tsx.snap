// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cell Component onClick and onContextMenu handlers called for active cells 1`] = `
<DocumentFragment>
  <div
    class="cell-1-1 css-1ywq03a"
    type="10"
  />
</DocumentFragment>
`;
